src/MessageRepository.php

```diff
@@ -26,4 +26,9 @@ public function retrieveAll(AggregateRootId $id): Generator;
      * @throws UnableToRetrieveMessages
      */
     public function retrieveAllAfterVersion(AggregateRootId $id, int $aggregateRootVersion): Generator;
+
+    /**
+     * @return Generator<int, Message, void, PaginationCursor>
+     */
+    public function paginate(int $perPage, PaginationCursor $cursor): Generator;
```

Comment:
I think it would make sense to shift `$perPage` into the PaginationCursor object, because then this method is very strongly typed and only requires a single input.

Reply:
The `@return` comment is less than ideal and is pretty weakly typed. The only way to determine if we are at the end of the stream is:

```php
$messages = $repository->paginate(100, $cursor);
/* foreach, or whatever */

// $next type cannot be inferred
$next = $messages->getReturn();

if ($next === null) {
    // we are probably at the end?
}
```

I think it would be helpful to introduce a `MessageCollection` or `MessagePage` class that wraps the generator and can provide more strongly typed information. Such a class would implement `IteratorAggregate` to expose the underlying generator, so the overhead of the layer would be very minimal.

Reply:
I also just realized that this method doesn't have, or allow for, any restriction by aggregate root ID. As such, it will replay every single message in the repository (within the constraints of the cursor). That would never work for our use case, as we only ever replay messages for a single aggregate root at a time.

The more I think about this, I am convinced that releasing this feature as it stands would lock EventSauce into a pattern that is inflexible and would cause regret at a later date.

Reply:
@frankdejonge that's kinda my modus operandi 😅 I'm not saying that the _only_ way to replay should be single-aggregate _or_ all-aggregate, but rather that binary is inflexible. For example, we actually extend the `MessageRepository` interface with a `public function findBy(MessageQuery $query): MessageCollection` method to allow for more precision when it comes to getting a selection a messages, for replay or other needs.

Reply:
Could you elaborate what in this design you'd mark as binary and inflexible? Always looking to improve, I just don't see what you're seeing.

Reply:
Well, the current implementation is a binary in that the methods available are:

```php
// Fetch ALL the messages for a SINGLE aggregate root.
public function retrieveAll(AggregateRootId $id): Generator;

// Fetch ALL the messages for ALL aggregates in the repository, with paging.
public function paginate(int $perPage, PaginationCursor $cursor): Generator;
```

There is no room in between here to do any of the following:

- fetch messages of a specific type
- fetch messages within a specific date range; see other comments about `int|string` typing for paging, which only provides one side of a start/stop range
- fetch messages for n+1 aggregate roots by ID
- ... or any other kind of criteria

Reply:
I think those use-cases are all rather custom, and therefore imho it is fine that you need to extend the message repository for that. I like the idea of moving `perPage` to the `offsetCursor` implementation (and rename it too limit). 

Allowing a custom paginate implementation depending on the store you are using, which is a more common use-case.